### PR TITLE
lastfm query tool

### PR DESCRIPTION
**Pull request by callesonne**
A basic script doing what was outlined in the "Create tools for querying the last.fm database" card on Trello. 

A summary of how the script accomplishes this:
get tags for any given set of track ids: it returns a dict linking track ids to its tags.
collect a count of tags for each track: again returns a dict linking track ids to the tag count.
collect a count for each tag for a set of track ids: returns dict linking tag to its count among track ids.
assign a threshold for number of tags and filter out tracks with less than this number: Takes in a list of track ids, and removes the ones with less than a certain amount of tags.

Now I do not know exactly how we will use this script so for now I was using dicts as return types, but of course this is not set in stone and depending on usage we might instead other return types. Changing the return type should require minimal modification.

### CODE REVIEW

**Comment by hcw1026** [comment]
REVIEW ON CALLE_LASTFM_QUERY
Very structured script!
But there are some minor problems:
1. Is there a way to make the names of the functions more systematic? as it was quite hard to follow... or can you make a summary (list) of all the functions? e.g. get_tag: tid--> tid_num  (small problem)

2. Some functions have only one direction. E.g. if I want to get tid from tid_num. 

3. the function get_tags_from_tids has an issue when inputting a string:
e.g. get_tags_from_tids(['TRQFQRL128F92E27F7']) works but get_tags_from_tids('TRQFQRL128F92E27F7') does not work
Would it be better if you add a check within the script to detect string or list? or at least throw an error if someone input a string by mistake?

4. tag_count(): in the else: count_dict[tag] = 1, is it wrong or am I wrong? I thought this will initialise the count_dict.

**Comment by callesonne** [comment]
Thank you for reviewing, very good points!
I am currently fixing 1, 2 and 3. As for 4:
Yes that is intended, it checks if tag is in the dictionary already as a key. If it is then add one to number of occurences, else initialize it to 1.

**Comment by callesonne** [comment]
I am still very unsure about the function names, now that I had to add even more they get confusing even for me which says something... If someone has suggested that would be greatly appreciated

**Comment by hcw1026** [comment]
I just have one more suggestion (not so important..)
do you want to add this line to the function:
If type(tids)!=list:
   tids = list(tids)

Also, should I merge it now?

**Comment by hcw1026** [comment]
I just discovered another issue:
I can't specify the path of the database unless I manually get into the code and change it, and I don't really know how we should solve it... First option: We mask it with a class? Second option: We manually change it every time

**Comment by callesonne** [comment]
> I just discovered another issue:
> I can't specify the path of the database unless I manually get into the code and change it, and I don't really know how we should solve it... First option: We mask it with a class? Second option: We manually change it every time

Good point, just realised the same thing today actually! What do you mean by making it a class? I was maybe thinking we could add a function set_tag_db_path or something similar.

**Comment by callesonne** [comment]
I am not sure now though, since having a function would force you to run it in every file that you want to query the tag database. Maybe it will be easier to just change it manually. Does your class solution circumvent this?

**Comment by callesonne** [comment]
> I just have one more suggestion (not so important..)
> do you want to add this line to the function:
> If type(tids)!=list:
> tids = list(tids)
> 
> Also, should I merge it now?

I am assuming you are talking about get_tags_dict and similar, I added a function get_tags which lets you just input a single tid as a string. I have skipped doing any error handling, since it is just us that will use the file and we know whats in it but If we were to use some error handling I would prefer to throw errors on anything that is not a list, forcing the user to convert it into a list before inputting it into the function. Otherwise the user might be unaware of this conversion happening.

**Comment by hcw1026** [comment]
Are you familiar with the structure of python class?
It is probability the easiest way to solve this issue but I guess is a more beautiful way?
Basically, you can make a class
Class last_fm_query( ):
  def __init__(self, path):
      self.path=path
      etc

  def youfunction(self, tid):
    etc

I don't know if this is the best solution tho.

**Comment by callesonne** [comment]
So you would suggest to put all functions inside of a class?

**Comment by callesonne** [comment]
Wouldnt this still leave the (minor) inconvenience of having to instantiate the class everytime you want to query the database?

**Comment by hcw1026** [comment]
Yeah this is true. Or you can put the path='the location in the server' as default' inside all the functions. And give an error if path is wrong? It's your script so you decide

**Comment by callesonne** [comment]
> Yeah this is true. Or you can put the path='the location in the server' as default' inside all the functions. And give an error if path is wrong? It's your script so you decide

This is probably the best solution.

**Comment by callesonne** [comment]
I am not sure now actually, I think I might go with a function actually. Because when I started changing some of the functions to use the path parameter, it became very ugly when you needed to chain calls, e.g. `tid_num_to_tag_num(tid_to_tid_num(tid, path=path), path=path)` Having to explicitly use it everytime makes it very inconvenient imo.

**Review by pyg4thr1x — COMMENTED** [review]
I really like how the script is structured, and I really like how you solved the path problem (had a similar issue...).

I made three (extremely minor) changes: 1) move one comment (please accept me as I am), 2) add a .gitignore, so that automatically generated files don't get in the way every time, 3) renamed 'tools' into 'modules', so that we can have a 'modules' folder for modules to be imported, and a 'scripts' folder for scripts to be run once in order to generate some output

**Comment by callesonne** [comment]
Is this pretty much done now? Or do you guys want me to do some error handling for the inputs as well (thinking simple type checking)?

**Review by pyg4thr1x — APPROVED** [review]
It's ok for me. We can still make minor changes after merging... @hcw1026? Are you happy with merging?
